R/utils.R

```diff
@@ -259,84 +259,333 @@ rowcol_to_dim <- function(row, col) {
+#'   can be coerced to data frame
 #' @examples
-#' # either vectors
+#' # Provide coordinates
+#' wb_dims()
+#' wb_dims(1, 4)
+#' wb_dims(rows = 1, cols = 4)
+#' wb_dims(start_row = 4)
+#' wb_dims(start_col = 2)
+#' wb_dims(1:4, 6:9, start_row = 5)
+#' # Provide  vectors
+#' wb_dims(1:10, 1:5)
 #' wb_dims(rows = 1:10, cols = 1:10)
+#' # provide `start_col` / `start_row`
+#' wb_dims(rows = 1:10, cols = 1:10, start_row = 2)
+#' wb_dims(rows = 1:10, cols = 1:10, start_col = 2)
 #' # or objects
-#' wb_dims(mtcars)
+#' wb_dims(x = mtcars)
+#' # dims of all the data of mtcars.
+#' wb_dims(x = mtcars, col_names = FALSE)
 #' @export
 wb_dims <- function(...) {
-
   args <- list(...)
-  nams <- names(args)
+  lengt <- length(args)
+  if (lengt == 0 || (lengt == 1 && is.null(args[[1]]))) {
+    return("A1")
+  }
 
-  col_names <- args$col_names
-  row_names <- args$row_names
+  # nams cannot be NULL now
+  nams <- names(args) %||% rep("", lengt)
+  valid_arg_nams <- c("x", "rows", "cols", "start_row", "start_col", "row_names", "col_names")
+  any_args_named <- any(nzchar(nams))
+  # unused, but can be used, if we need to check if any, but not all
+  # has_some_named_args <- any(!nzchar(nams)) & any(nzchar(nams))
+  # Check if valid args were provided if any argument is named.
+  if (any_args_named) {
+    match.arg_wrapper(arg = nams, choices = c(valid_arg_nams, ""), several.ok = TRUE, fn_name = "`wb_dims()`")
+  }
+  # After this point, no need to search for invalid arguments!
 
-  cnam_null <- is.null(col_names)
-  rnam_null <- is.null(row_names)
+  n_unnamed_args <- length(which(!nzchar(nams)))
+  all_args_unnamed <- n_unnamed_args == lengt
+  # argument dispatch / creation here.
+  # All names provided, happy :)
+  # Checking if valid names were provided.
 
-  srow <- args$start_row
-  scol <- args$start_col
+  if (n_unnamed_args > 2) {
+    stop("only `rows` and `cols` can be provided without names. You must name all other arguments.")
+  }
+  if (lengt == 1 && all_args_unnamed) {
+    stop(
+      "Specifying a single unnamed argument is not handled by `wb_dims()`",
+      "use `x`, `start_row`/ `start_col`. You can also use `dims = NULL`"
+    )
+  }
+  ok_if_arg1_unnamed <-
+    is.atomic(args[[1]]) | any(nams %in% c("rows", "cols"))
 
-  scol_null <- is.null(scol)
-  srow_null <- is.null(srow)
+  if (nams[1] == "" && !ok_if_arg1_unnamed) {
+    stop(
+      "The first argument must either be named or be a vector.",
+      "Providing a single named argument must either be `start_row`, `start_col` or `x`."
+    )
+  }
+  if (n_unnamed_args == 1 && lengt > 1 && !"rows" %in% nams) {
+    message("Assuming the first unnamed argument to be `rows`.")
+    nams[which(nams == "")[1]] <- "rows"
+    names(args) <- nams
+    n_unnamed_args <- length(which(!nzchar(nams)))
+    all_args_unnamed <- n_unnamed_args == lengt
+  }
+  if (n_unnamed_args == 1 && lengt > 1 && "rows" %in% nams) {
```

Comment:
why did you switch the logic here? Why not `!"cols" %in% nams`?

Reply:
I just want to safeguard in case someone messes up the logic.

if someone types,
```r
wb_dims(x = mtcars, rows = "vs")
```

The error message will redirect them to 
```r
wb_dims(x = mtcars, cols= "vs")
```

Reply:
ah okay. Makes sense, unless there's a rowname "vs"

Reply:
It currently does not support selecting rows, based on their row name. There could be advantages, but not sure if it's worth the effort...

Reply:
Fine by me. We do not have to make this overly complicated